IPNS.md

```diff
@@ -29,95 +38,302 @@ Each time a file is modified, its content address changes. As a consequence, the
+ optional bytes data = 9;
 }
 ```
 
-## Protocol
+Notes:
 
-Taking into consideration a p2p network, each peer should be able to publish IPNS records to the network, as well as to resolve the IPNS records published by other peers.
+- For legacy reasons, some values must be stored in both `IpnsEntry` protobuf and `IpnsEntry.data` CBOR.
+  This should not be ignored, as it impacts interoperability with old software.
 
-When a node intends to publish a record to the network, an IPNS record needs to be created first. The node needs to have a previously generated asymmetric key pair to create the record according to the datastructure previously specified. It is important pointing out that the record needs to be uniquely identified in the network. As a result, the record identifier should be a hash of the public key used to sign the record.
+### Record Size Limit
 
-As an IPNS record may be updated during its lifetime, a versioning related logic is needed during the publish process. As a consequence, the record must be stored locally, in order to enable the publisher to understand which is the most recent record published. Accordingly, before creating the record, the node must verify if a previous version of the record exists, and update the sequence value for the new record being created.
+IPNS implementations MUST support sending and receiving a serialized
+`IpnsEntry` less than or equal to **10 KiB** in size.
 
-Once the record is created, it is ready to be spread through the network. This way, a peer can use whatever routing system it supports to make the record accessible to the remaining peers of the network.
+Records over the limit MAY be ignored. Handling records larger than the
+limit is not recommended so as to keep compatibility with implementations and
+transports that follow this specification.
 
-On the other side, each peer must be able to get a record published by another node. It only needs to have the unique identifier used to publish the record to the network. Taking into account the routing system being used, we may obtain a set of occurrences of the record from the network. In this case, records can be compared using the sequence number, in order to obtain the most recent one.
+### Backward Compatibility
 
-As soon as the node has the most recent record, the signature and the validity must be verified, in order to conclude that the record is still valid and not compromised.
+Implementations that want to interop with the public IPFS swarm MUST maintain
+backward compatibility for legacy consumers of IPNS records:
 
-Finally, the network nodes may also republish their records, so that the records in the network continue to be valid to the other nodes.
 
-## Overview
+- A legacy publisher MUST always be able to update to the latest implementation
+  of this specification without breaking record resolution for legacy consumers.
+- A legacy consumer MUST always be able to resolve IPNS name, even when publisher
```

Comment:
does "resolve IPNS" name mean they should be able to validate the signature? Or just that they can read the record?
Isn't that already not the case since supporting all Key types isn't required there's already consumers that can't validate all signtures? And new key types (or signing prefixes) would invalidate old consumers?

Reply:
> does "resolve IPNS" name mean they should be able to validate the signature?

@paulgmiller  Yes, IPNS resolution includes signature verification step: records that have invalid signatureV2 are ignored by implementations like Kubo or js-ipfs.

> And new key types (or signing prefixes) would invalidate old consumers?

The use case we are protecting here is preexisting IPNS client resolving preexisting IPNS names for mission-critical updates. In that case, the key type will not change, records are published with the same key, and the old signature must be present next to the new one to ensure the name still resolves for the legacy client.

An example scenario is as follows:
-  IoT device implemented minimal support for RSA and V1, and uses it for polling for new firmware updates.
- It retrieves record with V1+V2 signatures, validates V1 and fetches new firmware
- The new firmware has support for IPNS V2, and uses a new IPNS name for polling for updates (requires V2 signatures, and can also use a different key type like Ed25519).